tiri/src/lib.rs

```diff
@@ -70,3 +73,33 @@ impl<'t> Interp<'t> {
         unimplemented!();
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::Interp;
+    use test::black_box;
+    use yktrace::{start_tracing, TracingKind};
+
+    // Some work to trace.
+    #[inline(never)]
+    fn work(x: usize, y: usize) -> usize {
+        let mut res = 0;
+        while res < y {
+            res += x;
+        }
+        res
+    }
+
+    #[test]
+    #[ignore] // FIXME -- This will fail until we investigate why SIR is missing.
+    fn interp_simple_trace() {
+        let tracer = start_tracing(Some(TracingKind::SoftwareTracing));
+        let res = black_box(work(black_box(3), black_box(13)));
```

Comment:
Surely there's no point in doing `black_box(3)` -- an integer can't realistically be optimised!

Reply:
It's to prevent constant propagation into the function body. Otherwise the body may be replaced by a value.

Reply:
Doesn't the outer `black_box` prevent that happening?

Reply:
Not necessarily. The addition could still be removed. Ultimately it's the arguments to the addition operator that we must prevent the compiler from knowing ahead of time.

The outer black box is actually probably not necessary.

Reply:
(It's probably sufficient to only guard only one argument with a black box, too, but having two won't hurt)

Reply:
We should understand what `black_box` does before we end up using it, otherwise we've no real idea if we're over or under using it. My guess -- but it is a guess! -- is that we've overused it here. But I might be wrong, and perhaps we need it in more places! We should understand it better IMHO.

Reply:
`black_box()` just inserts a chunk of inline assember code which is essentially the ID function:
https://doc.rust-lang.org/1.1.0/src/test/lib.rs.html#1054-1059

It stops optimisations crossing the black box.

My understanding, based on this, is that If you guard an argument to our work function with a BB, then operations that use the argument can't be replaced with a value.

If we guard the whole call, but not the args, I'd expect the call to be uninlinable, since the compiler loses knowledge about the return value, but not the computation in the leadup.

I can investigate further, or we can drop the outer black box.

Reply:
The docs say:

> A function that is opaque to the optimizer, to allow benchmarks to pretend to use outputs to assist in avoiding dead-code elimination.

which feels to me to be somewhat different than what you've said?

Reply:
Yeah, it'd prevent DCE, but I'm very sure it'd block other optimisations. Let me try and build a proof.

Reply:
Here's a case where we clearly see the whole call to the function `myadd` being replaced with a constant folded value `999`:

![image](https://user-images.githubusercontent.com/604955/61135445-88a60f00-a4b9-11e9-8903-a5e215302b8f.png)

And here's the same with the arguments black boxed:
![image](https://user-images.githubusercontent.com/604955/61135379-5eece800-a4b9-11e9-8779-a34105d661d8.png)

Notice here that the compiler is unable to jump to the conclusion `999`, as it doesn't know what the inline assembler code does. It is forced to perform the addition.

[As an aside, notice that the function is still inlined regardless of the black box]

Makes sense?

Reply:
What happens if you put the `black_box` around the outer function in the second case?

Reply:
I suspect the call will be replaced with `999` again.

[black_boxes around entire function calls are useful for something else though: keeping code from re-ordering]

One moment...

Reply:
As expected, the compiler performs constant propagation:
![image](https://user-images.githubusercontent.com/604955/61139154-b5115980-a4c0-11e9-94d4-1304b2a0262c.png)

Reply:
That's very interesting! `black_box` does not work as I expected, which is probably my fault.

OK, so going back to this PR ;) I now think we should drop the outer `black_box` call since it doesn't do anything useful in this context?

Reply:
Yep, I agree.

Reply:
You've also highlighted that the Rust `black_box` docs could be improved. I might do a PR against upstream.

Reply:
Cool!